### PR TITLE
[Feature] Support CopyPaste for RotatedBoxes

### LINKED ISSUE #82: maybe you can add the copy paste method in the data augmentation
**Describe the feature**

**Motivation**
A clear and concise description of the motivation of the feature.
Ex1. It is inconvenient when [....].
Ex2. There is a recent paper [....], which is very helpful for [....].

**Related resources**
If there is an official code release or third-party implementations, please also provide the information here, which would be very helpful.

**Additional context**
If you would like to implement the feature and create a PR, please leave a comment here and that would be much appreciated.

### PR DESCRIPTION
**Pull request by nijkah**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Closes #82 

## Modification

1. Add `RBitMasks`
2. Add `RCopyPaste`

Welcome any suggestions on the current implementation!

## Example
dota_1

dota_2

dota_3

### CODE REVIEW

File: mmrotate/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -544,3 +547,141 @@ def _filter_box_candidates(self, bboxes, labels, w, h):
...
+    def __init__(
+        self,
+        max_num_pasted=100,
+        rbbox_occluded_iou_thr=0.3,
+        mask_occluded_thr=300,
+        selected=True,
+        version='le90',
+    ):
+        self.max_num_pasted = max_num_pasted
+        self.rbbox_iou_thr = rbbox_occluded_iou_thr
```
Comment (nijkah) [inline mmrotate/datasets/pipelines/transforms.py:594]:
```suggestion
        self.rbbox_occluded_iou_thr= rbbox_occluded_iou_thr
```

File: mmrotate/core/mask/structures.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        y_any = self.masks.any(axis=2)
+        for idx in range(num_masks):
+            x = np.where(x_any[idx, :])[0]
+            y = np.where(y_any[idx, :])[0]
+            if len(x) > 0 and len(y) > 0:
+                contours = cv2.findContours(self.masks[idx], cv2.RETR_TREE,
+                                            cv2.CHAIN_APPROX_SIMPLE)[0][0]
+                (cx, cy), (w, h), a = cv2.minAreaRect(contours)
+                rboxes[idx, :] = np.array(
+                    [cx, cy, w, h, np.radians(a)], dtype=np.float32)
```
Comment (nijkah) [inline mmrotate/core/mask/structures.py:32]: Here, we need to support `mmrotate.core.bbox.poly2obb` for various angle versions.
However, `poly2obb` supposes that polygons have 4 vertices.
How can we support this?

**Comment by zytx121** [comment]
Excellent job! It's very useful.

File: mmrotate/core/mask/structures.py (hunk 2 above)
Comment (zytx121) [reply mmrotate/core/mask/structures.py:32]: In mmrotate1.x, we supposes all rotated boxes use `oc` angle version during data transforms. 
We only need to distinguish different angle versions within the box_head.
So I think it's OK to write like this.

**Comment by zytx121** [comment]
Sorry to reply so late. Anyway, I'm back.

The project structure of MMDet and mmrotate has been updated. Please update your code according to https://github.com/open-mmlab/mmdetection/blob/3.x/mmdet/structures/mask/structures.py

I will continue to pay attention to this PR and hope your enthusiasm will not fade.

**Comment by hongxrsysu** [comment]
I use this code, and met a bug:

>   File "tools/misc/browse_dataset.py", line 110, in <module>
    main()
  File "tools/misc/browse_dataset.py", line 86, in main
    for item in dataset:
  File "/root/miniconda3/lib/python3.8/site-packages/mmdet/datasets/dataset_wrappers.py", line 431, in __getitem__
    updated_results = transform(copy.deepcopy(results))
  File "/root/miniconda3/lib/python3.8/site-packages/mmdet/datasets/pipelines/transforms.py", line 2830, in __call__
    selected_results = self._select_object(results['mix_results'][0])
  File "/root/miniconda3/lib/python3.8/site-packages/mmdet/datasets/pipelines/transforms.py", line 2839, in _select_object
    masks = results['gt_masks']
KeyError: 'gt_masks'

I find that "gt_masks" is in the "mmrotate/datasets/pipelines/transforms.py". 
image


So how to fix it?

**Comment by nijkah** [comment]
@hongxrsysu I think it works normally after upgrading mmdet >= 2.26.0.
It supposes that `CopyPaste` supports generating a pseudo `gt_masks` when `gt_masks` is not available.


Refer these links.
https://github.com/open-mmlab/mmdetection/releases/tag/v2.26.0
https://github.com/open-mmlab/mmdetection/commit/1f8b195f723d2bcd3b8c79b62b0bc1365f2fa290

**Comment by nijkah** [comment]
@zytx121 The `CopyPaste` implementation seems quite different between `mmdet==2.x` and `mmdet==3.x`.
For example, in `mmdet==3.x`, it does not support generating pseudo masks from bounding boxes.

Is there any good way to support both `mmrotate==1.x` and `mmrotate==0.x`?
Or would it be okay to implement each in a different branch?

**Comment by hongxrsysu** [comment]
> @hongxrsysu I think it works normally after upgrading mmdet >= 2.26.0. It supposes that `CopyPaste` supports generating a pseudo `gt_masks` when `gt_masks` is not available.
> 
> Please refer to these links. https://github.com/open-mmlab/mmdetection/releases/tag/v2.26.0 [open-mmlab/mmdetection#8905](https://github.com/open-mmlab/mmdetection/pull/8905)

Thanks! It works!

**Comment by zytx121** [comment]
> @zytx121 The `CopyPaste` implementation seems quite different between `mmdet==2.x` and `mmdet==3.x`. For example, in `mmdet==3.x`, it does not support generating pseudo masks from bounding boxes.
> 
> Is there any good way to support both `mmrotate==1.x` and `mmrotate==0.x`? Or would it be okay to implement each in a different branch?

As far as I know, there is currently no way to support two branches at the same time. It is okay to implement each in a different branch

**Comment by nijkah** [comment]
Hi, @zytx121. Apologize for the slow progress.
Recently. I create a PR for MMRotate=1.x. 😄 

I'll close this PR.

**Pull request closed without merging**